File: Assets/Game/Scripts/Moduls/P_Modul_01.lua
local ObjectStateBackupUtils = Ballance2.Utils.ObjectStateBackupUtils

---P_Modul_01
---栅栏机关
---@class P_Modul_01 : ModulBase
---@field P_Modul_01_Rinne PhysicsObject
---@field P_Modul_01_Filter PhysicsObject
---@field P_Modul_01_Pusher PhysicsObject
P_Modul_01 = ModulBase:extend()

function P_Modul_01:new()
  P_Modul_01.super.new(self)
  self.EnableBallRangeChecker = true
  self.BallCheckeRange = 60
end
function P_Modul_01:Start()
  if not self.IsPreviewMode then
    self.P_Modul_01_Pusher.CollisionID = GamePlay.BallSoundManager:GetSoundCollIDByName('WoodOnlyHit')
  end
end
function P_Modul_01:Active()
  self.gameObject:SetActive(true)
  self.P_Modul_01_Rinne:Physicalize()
  self.P_Modul_01_Filter:Physicalize()
  self.P_Modul_01_Pusher:Physicalize()
end
function P_Modul_01:Deactive()
  self.P_Modul_01_Rinne:UnPhysicalize(true)
  self.P_Modul_01_Filter:UnPhysicalize(true)
  self.P_Modul_01_Pusher:UnPhysicalize(true)
  self.gameObject:SetActive(false)
end
function P_Modul_01:Reset()
  ObjectStateBackupUtils.RestoreObject(self.gameObject)
end
function P_Modul_01:Backup()
  ObjectStateBackupUtils.BackUpObject(self.gameObject)
end

function P_Modul_01:ActiveForPreview()
  self.gameObject:SetActive(true)
end
function P_Modul_01:DeactiveForPreview()
  self.gameObject:SetActive(false)
end

function P_Modul_01:BallEnterRange()
  if not self.IsPreviewMode and self.IsActive then
    self.P_Modul_01_Pusher:WakeUp()
  end
end

function CreateClass:P_Modul_01()
  return P_Modul_01()
end

File: Assets/Game/GamePlay/Scripts/Moduls/P_Modul_01.cs
using Ballance2.Utils;
using BallancePhysics.Wapper;

namespace Ballance2.Game.GamePlay.Moduls
{
  /// <summary>
  /// 栅栏机关
  /// </summary>
  public class P_Modul_01 : ModulBase 
  {
    public PhysicsObject P_Modul_01_Rinne;
    public PhysicsObject P_Modul_01_Filter;
    public PhysicsObject P_Modul_01_Pusher;

    protected override void Start()
    {
      if (!IsPreviewMode) 
        P_Modul_01_Pusher.CollisionID = GamePlayManager.Instance.BallSoundManager.GetSoundCollIDByName("WoodOnlyHit");
      base.Start();
    }

    public override void Active()
    {
      gameObject.SetActive(true);
      P_Modul_01_Rinne.gameObject.SetActive(true);
      P_Modul_01_Filter.gameObject.SetActive(true);
      P_Modul_01_Pusher.gameObject.SetActive(true);
      P_Modul_01_Rinne.Physicalize();
      P_Modul_01_Filter.Physicalize();
      P_Modul_01_Pusher.Physicalize();
    }
    public override void Deactive()
    {
      P_Modul_01_Rinne.UnPhysicalize(true);
      P_Modul_01_Filter.UnPhysicalize(true);
      P_Modul_01_Pusher.UnPhysicalize(true);
      gameObject.SetActive(false);
    }

    public override void ActiveForPreview()
    {
      gameObject.SetActive(true);
    }
    public override void DeactiveForPreview()
    {
      gameObject.SetActive(false);
    }

    public override void Reset(ModulBaseResetType type)
    {
      ObjectStateBackupUtils.RestoreObjectAndChilds(gameObject);
    }
    public override void Backup()
    {
      ObjectStateBackupUtils.BackUpObjectAndChilds(gameObject);
    }
  }
}